### PR TITLE
OBA #496 Preferred redirect header

### PR DESCRIPTION
**Pull request by elroman**

### CODE REVIEW

File: opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml [hunk 1]
```diff
@@ -75,14 +78,16 @@
...
-            <column header="aspsp_sca_approaches" name="sca_approaches" type="STRING"/>
+            <column header="uuid"                   name="bank_uuid"                type="STRING"/>
+            <column header="name"                   name="name"                     type="skip"/>
+            <column header="bic"                    name="bic"                      type="skip"/>
+            <column header="url"                    name="url"                      type="STRING"/>
+            <column header="adapter_id"             name="adapter_id"               type="STRING"/>
+            <column header="bank_code"              name="bank_code"                type="skip"/>
+            <column header="idp_url"                name="idp_url"                  type="STRING"/>
+            <column header="aspsp_sca_approaches"   name="sca_approaches"           type="STRING"/>
+            <column header="preferred_approach"     name="preferred_approach"       type="STRING"/>
```
Comment (valb3r) [inline opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml:89]: Please avoid changing CSV structure. It is fine to add rows, but not columns. This file technically belongs to other project

File: opba-facade-protocol-api-shared/src/main/java/de/adorsys/opba/protocol/api/common/CurrentBankProfile.java [hunk 2]
```diff
@@ -36,4 +36,9 @@
      * Preferred SCA approach for this ASPSP.
      */
     Approach getPreferredApproach();
+
+    /**
+     * ASPSP forces preferred SCA approach.
+     */
+    Boolean getForcePreferredApproach();
```
Comment (valb3r) [inline opba-facade-protocol-api-shared/src/main/java/de/adorsys/opba/protocol/api/common/CurrentBankProfile.java:43]: Why do we have it as a nullable type - we can only force approach by sending header or not. Suggest to use as the non-null type

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/dto/Xs2aStandardHeaders.java [hunk 3]
```diff
@@ -45,12 +48,21 @@
...
+    @Nullable
+    private Boolean tppRedirectPreferred;
+
     public RequestHeaders toHeaders() {
         Map<String, String> allValues = new HashMap<>();
 
         allValues.put(PSU_ID, psuId);
         allValues.put(X_REQUEST_ID, requestId);
         allValues.put(X_GTW_ASPSP_ID, aspspId);
+        Optional.ofNullable(tppRedirectPreferred)
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/dto/Xs2aStandardHeaders.java:64]:
```
if (null != tppRedirectPreferred) {
allValues.put(TPP_REDIRECT_PREFERRED, String.valueOf(tppRedirectPreferred));
}
```
Seem to be easier to understand

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/authenticate/TppRedirectPreferredResolver.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import de.adorsys.opba.protocol.api.common.Approach;
+import de.adorsys.opba.protocol.api.common.CurrentBankProfile;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TppRedirectPreferredResolver {
+
+    public Boolean computeRedirectPreferred(CurrentBankProfile config) {
+        return Boolean.TRUE.equals(config.getForcePreferredApproach())
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/authenticate/TppRedirectPreferredResolver.java:11]: Can be simplified to
```
return config.getForcePreferredApproach() ? config.getPreferredApproach() == Approach.REDIRECT : null;
```

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/authenticate/TppRedirectPreferredResolver.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package de.adorsys.opba.protocol.xs2a.service.xs2a.authenticate;
+
+import de.adorsys.opba.protocol.api.common.Approach;
+import de.adorsys.opba.protocol.api.common.CurrentBankProfile;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TppRedirectPreferredResolver {
+
+    public Boolean computeRedirectPreferred(CurrentBankProfile config) {
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/authenticate/TppRedirectPreferredResolver.java:10]: `isRedirectApproachPreferred` seem to be more meaningful name

File: opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml [hunk 6]
```diff
@@ -37,6 +37,7 @@
                 <constraints unique="true"/>
             </column>
             <column name="preferred_approach" type="VARCHAR(32)"/>
+            <column name="force_preferred_approach" type="boolean" defaultValue="false"/>
```
Comment (valb3r) [inline opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml:40]: Given the PSD2 logic 'force_preferred_approach' name looks misleading. We can't force EMBEDDED mode unfortunately. 'try_to_use_preferred_approach' seem a better name

Comment (valb3r) [reply opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml:40]: Also, doesn't seem that this is nullable column

File: opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml [hunk 7]
```diff
@@ -119,7 +124,7 @@
 
     <changeSet id="fill_services" author="mhr">
         <sql dbms="postgresql" stripComments="true">
-            UPDATE opb_bank_profile SET preferred_approach = 'REDIRECT' WHERE true ;
+            UPDATE opb_bank_profile SET preferred_approach = 'REDIRECT' WHERE preferred_approach = '' ;
```
Comment (valb3r) [inline opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml:127]: Please keep the previous statement for now

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/authenticate/TppRedirectPreferredResolver.java (hunk 4 above)
Comment (elroman) [reply opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/authenticate/TppRedirectPreferredResolver.java]: done

File: opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml (hunk 7 above)
Comment (elroman) [reply opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml]: done

File: opba-facade-protocol-api-shared/src/main/java/de/adorsys/opba/protocol/api/common/CurrentBankProfile.java (hunk 2 above)
Comment (elroman) [reply opba-facade-protocol-api-shared/src/main/java/de/adorsys/opba/protocol/api/common/CurrentBankProfile.java]: done

File: opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml (hunk 6 above)
Comment (elroman) [reply opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml]: done

**Pull request merged**